rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,98 @@
+| MessagePack       | 135    | 222        | 237      | 119      | 296          | 187         |
+| **Compiled Libs** |        |            |          |          |              |             |
+| Flatbuffers       | 1046   | 14706      | 250000   | 1065     | 264          | 181         |
+| Bincode           | 5882   | 8772       | 9524     | 4016     | 163          | 129         |
+| Protobuf          | 859    | 1140       | 1163     | 480      | 154          | 141         |
+| Prost             | 1225   | 1866       | 1984     | 962      | 154          | 142         |
+
+Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks). Some of these formats (JSON for example) still leaves us with a marginal concern around dropped metadata.
+
+While it is tempting to look at perf benchmarks and to just pick the best performing example, the practical requirements around the problem for us might give us some limitations. We have a few specific domains of concern around the chosen path:
+
+- Performance
+- Compatibility with Transports
+- Maintaining backcompat
+- Tooling and Maintenance costs
+
+These four concerns don't fully enumerate the problem but I think they're a really excellent distillation of what we need to keep in mind. So, to reiterate - our ideal solution solves for all four but that might just not be possible.
+
+It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MessagePack or CBOR. And their maintenance cost has already been paid since we're using them today. Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.
+
+With regards to the tansport and the suggested path of implementing HTTP/1.1 _before_ HTTP/2 we have to keep both perf and kubernetes issues in mind. [TCP causes us some problems in our K8s integration today](https://github.com/timberio/vector/issues/2070) and unfortunately this problem has quite a bit more context to it than can sanely be shared in this RFC. Suffice to say that any choice we make here has repurcussions on our deployment architecture and ultimately HTTP/2 with Protobufs provides us what we think are the right tradeoffs for our implementation. However, just quickly getting out a low-effort HTTP/1.1 implementation with batching solves some immediate pain for ourselves and our users around dynamic IPs and kubernetes load-balancing. I'd classify it as low-hanging fruit while a more thorough HTTP/2 and optimized Protobufs effort gets underway.
```

Comment:
Just noting that it is highly likely that the mechanisms we'd build to properly deploy the HTTP/2 should work for other protocols as well. We should land them as reusable bits.

Another thing - we have `http` source and `http` sink, that work with HTTP/1.1 already, but are, apparently, incompatible, as the sink does batching, while the source doesn't understand it.